### DISCUSSION TITLE
How can I set the Jupyter path?

### DISCUSSION THREAD

**Discussion by dlfivefifty** [quarto-dev/quarto-cli#9432] [Q&A] [answered]
### Description

Is it possible to set the Jupyter path directly when running Quarto in VSCode?

My setup seems to not be working otherwise. When I try to compile a document I get the message:
```
ERROR: Jupyter kernel 'julia-1.10' not found. Known kernels: python3. Run 'quarto check jupyter' with your python environment activated to check python version used.
```
So then I run `quarto check jupyter` and it says:
```
[✓] Checking Python 3 installation....OK
      Version: 3.12.3
      Path: /opt/homebrew/opt/python@3.12/bin/python3.12
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with python3 -m pip install jupyter
```
So then I run `python3 -m pip install jupyter` and it says:
```
× This environment is externally managed
╰─> To install Python packages system-wide, try brew install
```
So I run `brew install jupyter` but that doesn't resolve anything!

**Comment by mcanouil** [comment]
See <https://quarto.org/docs/computations/julia.html#installation>.

And regarding Jupyter kernels: https://docs.jupyter.org/en/latest/install/kernels.html

FYI, in the 1.5 pre-release there is native Julia engine: https://quarto.org/docs/computations/julia.html#using-the-julia-engine

**Comment by dlfivefifty** [comment, marked as answer]
In the end I fixed it by pointing quarto to the python installed by Conda.jl in Julia, not the homebrew version.  At first I did this by setting my PATH, but somehow when run in VSCode it was setting the `QUARTO_PYTHON` environment variable to still point to the homebrew version! Very annoying and its unclear whether it was the VSCode extension or something else setting the variable.

Solution: In my .zprofile I added a line `export QUARTO_PYTHON=${HOME}/.julia/conda/3/aarch64/bin/python3`. This now finds the right Jupyter.

**Reply by mcanouil** [reply]
The variable you described is in the documentation, but Quarto should not needs that.
Following the instructions I shared should work.
You could also double check your VSCode settings in case you changed the default Python interpreter path.

**Reply by dlfivefifty** [reply]
No those instructions didn't do anything since something was assigning `QUARTO_PYTHON` to point to a broken version of python.

Why would changing the default Python path also assign the `QUARTO_PYTHON` environment variable? Also something was modifying my Path to add Quarto. It's very clearly something in the quarto extension messing with my setup.

**Reply by mcanouil** [reply]
The extension uses your other extensions and your environment to set this variable.
So, it is very likely your VSCode setup contains non default or wrong values which confuses the Quarto extension.
As said, you should check your Python VSCode extension, in particular the I interpreter path.

By the way, I am on Mac, using Homebrew, and VSCode. I never faced any issues, so the issue lies in your setup.

**Reply by dlfivefifty** [reply]
>  I never faced any issues, so the issue lies in your setup.

Duh! That's why I'm asking for help to figure out what's wrong!

**Reply by mcanouil** [reply]
Then did you check your Python interpreter path in VSCode? Nearly all issues related to this are either a wrongly activated environment or a non default interpreter path set by the user.
Because, you never said anything about it and went straight talking about forcing Python path using `QUARTO_PYTHON` (https://quarto.org/docs/projects/environment.html).

**Reply by dlfivefifty** [reply]
I did and its just the default:

<img width="683" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/1229737/772e3c86-7499-49e1-8b24-9a96741f6dc4">

In a regular terminal I have:
```julia
solver@IC-NTVX7P67VQ ~ % which python
/Users/solver/.julia/conda/3/aarch64/bin/python
```

> you never said anything about it and went straight talking about forcing Python path using QUARTO_PYTHON

I didn't go straight to forcing! I did reasonable checks! I do not know why something was setting the QUARTO_PYTHON, it's generally not helpful to accuse people of not doing things when you don't know whether or not they did them.

**Reply by mcanouil** [reply]
You come for help, I asked to check some settings and you only replied that you forced Python path using the environment variable.

And when I asked again to check your reply was:

> Duh! That's why I'm asking for help to figure out what's wrong!

We appreciate your participation, but it’s important to note that mutual cooperation is essential. To ensure you receive the best possible help, we kindly ask that you provide clear information and that you reply to the requests made to help you.

Constructive feedback is always welcome as it contributes positively to our discussion, which unfortunately was not the case of the one quoted above for example.

**Reply by dlfivefifty** [reply]
Mate, I asked a specific question about setting the Jupyter path and you just sent links to the docs, none of which said how to set the Jupyter path. I managed to work around the issue which I shared with others and you posted a snarky comment about there being something wrong with my setup. Then you accused me of not even checking the python path. So I don’t see anything you’ve done in this conversation that has been at all “constructive”.

**Reply by mcanouil** [reply]
Firstly, I’d like to clarify that I prefer not to be addressed as “mate,” as we are not acquainted. I apologise if my previous messages came across as unhelpful or dismissive; that was not my intention.
Regarding the Jupyter path issue, my goal was to direct you to comprehensive resources and checks that could potentially address a variety of related problems, including the one you were facing. I understand now that you were looking for a specific temporary workaround rather than a solution to the underlying issue.

**Reply by dlfivefifty** [reply]
I’ve been working on open-source software for a very long time, over 10 years. Often you get enquiries from people from different cultures who have different ways of expressing themselves. It’s important to give people the benefit of the doubt and not get offended by the way people express themselves or try to impose your own culture or way of expressing things on others. In this case “Mate” is just a standard term used to refer to others in a collegial way in probably half the English speaking world.

I appreciate that you are trying your best to help users but sometimes it’s best to focus on what they are actually asking. Or if you don’t know the answer it’s ok just to not respond, I understand this isn’t necessarily your job so sometimes a non-response is an answer in itself.

Despite the long thread I still don’t have an answer for how to set the Jupyter path manually. I’m assuming this means the answer is “you can’t”.

 (I also had a few other questions about setting the Julia environment but I’m guessing that’s not supported yet either since it’s not in the docs, so I think I’ll just stick to what’s working.)

**Reply by mcanouil** [reply]
There are no "Jupyter path".
Quarto uses the Jupyter install from your Python environment.
You should not have to do anything, but you can brute force the Python path (thus the "Jupyter path") via `QUARTO_PYTHON` which was in the link I shared.

Now, you can also try the 1.5 pre-release and use the new native Julia engine.

> I’ve been working on open-source software for a very long time, over 10 years. Often you get enquiries from people from different cultures who have different ways of expressing themselves. It’s important to give people the benefit of the doubt and not get offended by the way people express themselves or try to impose your own culture or way of expressing things on others

The benefit of the doubt goes both ways. I am not a native English speaker. So your assumption of me being offended or accusing you is your very own point of view.

The fact is that you never replied or wrote that you checked the VSCode setting I talked about, only after the third time.
You might think it was obvious but what is obvious to you might not be to others and from all previous discussions about very close or identical issues, this setting was amongst the sources of the issues.

**Reply by cscheid** [reply]
> I appreciate that you are trying your best to help users but sometimes it’s best to focus on what they are actually asking.

@dlfivefifty: these kinds of statements that imply that the responder isn't being appropriate are not constructive nor germane to your question. Please refrain from doing so in the future.

**Reply by dlfivefifty** [reply]
I don’t understand, I didn’t imply the responder was not being “appropriate”. I only meant that they weren’t addressing my question, and I felt they were being dismissive.

Again people come from very different backgrounds in only communication, and many people are neurodiverse (including myself) in a way that impacts their ability to communicate. I suggest you refrain from inferring disrespect in someone’s comments because it’s probably just someone who struggles with communication.

**Reply by dlfivefifty** [reply]
I would also like to not continue  this conversation please. I’m finding it upsetting and not productive. I’m sorry if my way of communicating has caused anyone distress, it was not my intention.

**Reply by cscheid** [reply]
This conversation is no longer productive, and this is the last message I will post here. The upshot is that:

- I have told you that your statement is not up to the standards we expect forum users to engage with people who are literally paid to help you here
- I'm currently the main developer of Quarto, and it's my call to make when people violate the [code of conduct](https://github.com/quarto-dev/quarto-cli?tab=coc-ov-file#readme). I pointed out how and why you have done so.
- Please refrain from doing that again in the future.
